### ISSUE TITLE
After load all, ?"?" gives error

### ISSUE THREAD

**Issue by wch** [hadley/devtools#574]
In RStudio, it shows:

```
Error in as.name(topic) : attempt to use zero-length variable name
```

One difference is that, in a clean session you get this:

```
> str(utils::`?`("?"))
Class 'help_files_with_topic'  atomic [1:1] /Library/Frameworks/R.framework/Versions/3.1/Resources/library/utils/help/Question
  ..- attr(*, "call")= language help(topic = "?", package = NULL)
  ..- attr(*, "topic")= chr "?"
  ..- attr(*, "tried_all_packages")= logi FALSE
  ..- attr(*, "type")= chr "html"
```

And after a `load_all()`:

```
> devtools::load_all(".")
> str(utils::`?`("?"))
Class 'help_files_with_topic'  atomic [1:2] /Users/winston/R/3.1/devtools/help/help /Library/Frameworks/R.framework/Versions/3.1/Resources/library/utils/help/Question
  ..- attr(*, "call")= language help(topic = "?", package = NULL)
  ..- attr(*, "topic")= chr "?"
  ..- attr(*, "tried_all_packages")= logi FALSE
  ..- attr(*, "type")= chr "html"
```

**Comment by hadley** [comment]
I can't reproduce this problem any more.

**Comment by wch** [comment]
In RStudio, If I `load_all` a different package (not devtools) and then do `?"?"`, nothing comes up in the RStudio help pane.


But if I do the same thing in a terminal, it seems to work:

```
> devtools::load_all()
Loading shiny
> ?"?"
Help on topic ‘?’ was found in the following packages:

  Package               Library
  devtools              /home/winston/R/3.1
  utils                 /usr/lib/R/library

Choose one 

1: Drop-in replacements for help and ? functions {devtools}
2: Documentation Shortcuts {utils}

Selection: 
```

**Comment by wch** [comment]
Looks like it's not just in RStudio. It also happens when you use web help from the terminal.

```
options(help_type = "html")
?"?"
```

This results in a browser going to http://127.0.0.1:10178/library/NULL/help/? , and it displays:

```
Error in as.name(topic) : attempt to use zero-length variable name
```

**Comment by wch** [comment]
With ``debug(`?`)`` and ``debug(utils::`?`)``, I found that it eventually calls `help("?", package = NULL)`.

If you run that command in a clean R session without loading a package, the help page shows up just fine.

But if you do `options(help_type = "html")` and `load_all()` on a (non-devtools) package, then when you run `help("?", package = NULL)`, you get that error. Also, `utils::help("?")` gives the same error. However, `utils::help("?", package = "utils")` gives the correct help page.

**Comment by wch** [comment]
In a clean R session with HTML help on, the URL for `?"?"` is http://127.0.0.1:15456/library/utils/html/Question.html.

But after doing a `load_all()`, the URL is http://127.0.0.1:10178/library/NULL/help/? , and I don't think this is a valid help URL.

The different URL is because it goes down this code path: https://github.com/wch/r-source/blob/5c1b07c22630ad725716a8d6bd821e3a5c7d1e42/src/library/utils/R/help.R#L153-L158

I think the only reasonable solution is to special-case the devtools shim_help so that it recognizes when `?"?"` is called and in that case does `help("?", package = "utils")`.

**Comment by wch** [comment]
The special casing would involve adding something like this near the end of `shim_help`:

```R
  else if (identical(e1_str, "?")) {
    utils::help("?", package = "utils")
  }
```

**Comment by wch** [comment]
A little more info: the real problem is that is that the URL isn't being escaped properly by `utils:::print.help_files_with_topic`. Instead of `http://127.0.0.1:13958/library/NULL/help/%3F`, it's going to `http://127.0.0.1:13958/library/NULL/help/?`.

Sort of fixing `utils:::print.help_files_with_topic`, it's possible to modify the help object so that it has the correct topic for html help, by escaping the `?`:

```R
# load a non-devtools package
load_all('../shiny')

options(help_type = "html")
z <- help("?", package = NULL)
str(z)
# Class 'help_files_with_topic'  atomic [1:2] /Users/winston/R/3.1/devtools/help/help /Library/Frameworks/R.framework/Versions/3.1/Resources/library/utils/help/Question
#   ..- attr(*, "call")= language utils::help(topic = `?`, package = NULL)
#   ..- attr(*, "topic")= chr "?"
#   ..- attr(*, "tried_all_packages")= logi FALSE
#   ..- attr(*, "type")= chr "html"

if (attr(z, "type") == "html") {
  # escape the ?
  attr(z, "topic") <- "%3F"
}
z
```

**Comment by wch** [comment]
Filed a bug report for R: https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=16173

**Comment by hadley** [comment]
Nice work! (BTW you're invisible on chat again)

**Comment by wch** [comment]
Looks like it's fixed in R-devel:  wch/r-source@3bf09244e8b422ab58a3413cf8736686d296b554
